### PR TITLE
Add snapshot validation webhook chart

### PR DESCRIPTION
**Pull request by masteryyh**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

We need to add snapshot validation webhook according to official documentation.

**Solution:**

Add snapshot validation webhook helm chart according to [here](https://github.com/kubernetes-csi/external-snapshotter/tree/master/deploy/kubernetes/webhook-example).

**Related Issue:**
https://github.com/harvester/harvester/issues/2983

**Test plan:**
1. Build an ISO and deploy using this branch;
2. After the cluster is ready, try to apply these two invalid snapshot, none of them should be applied and webhook should return an error message:
[invalid-snapshot-v1.yaml](https://github.com/kubernetes-csi/external-snapshotter/blob/021fad1cfe0277dd4052a4efe3f5345b6399d077/examples/kubernetes/invalid-snapshot-v1.yaml)
[invalid-snapshot-v1beta1.yaml](https://github.com/kubernetes-csi/external-snapshotter/blob/021fad1cfe0277dd4052a4efe3f5345b6399d077/examples/kubernetes/invalid-snapshot-v1beta1.yaml)

### CODE REVIEW

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v2
+name: snapshot-validation-webhook
+description: A Helm chart for Kubernetes
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/Chart.yaml]: typo: A Helm chart for CSI snapshot webhook validation

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  - name: "validation-webhook.groupsnapshot.storage.k8s.io"
+    rules:
+      - apiGroups:   ["groupsnapshot.storage.k8s.io"]
+        apiVersions: ["v1alpha1"]
+        operations:  ["CREATE", "UPDATE"]
+        resources:   ["volumegroupsnapshots", "volumegroupsnapshotcontents", "volumegroupsnapshotclasses"]
+        scope:       "*"
+    clientConfig:
+      service:
+        namespace: "default"
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml]: should this be `kube-system`?

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        name: "snapshot-validation-service"
+        path: "/volumegroupsnapshot"
+      caBundle: {{ $ca.Cert | b64enc }}
+    admissionReviewVersions: ["v1"]
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml]: should we also include `v1beta1` for backward compatibility support?

**Review by futuretea — APPROVED** [review]
LGTM, thanks.

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: snapshot-webhook-runner
+rules:
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotclasses"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["groupsnapshot.storage.k8s.io"]
+    resources: ["volumegroupsnapshotclasses"]
+    verbs: ["get", "list", "watch"]
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/rbac.yaml]: why is this needed?

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+---
+{{ $ca := genCA "snapshot-validation-webhook-ca" 365 }}
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml]: let's define it as 3650, otherwise certificate rotation will quickly become a problem.

**Review by guangbochen — COMMENTED** [review]
Additionaly, u don't need to commit other dependency charts if they are not modified.

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: "validation-webhook.snapshot.storage.k8s.io"
+webhooks:
+  - name: "validation-webhook.snapshot.storage.k8s.io"
+    rules:
+      - apiGroups:   ["snapshot.storage.k8s.io"]
+        apiVersions: ["v1", "v1beta1"]
+        operations:  ["CREATE", "UPDATE"]
+        resources:   ["volumesnapshots", "volumesnapshotcontents", "volumesnapshotclasses"]
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml]: remove the `volumesnapshotclasses` to be consistent with RBAC

File: deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+{{ $key := b64enc $cert.Key }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: snapshot-validation-webhook-ca
+  namespace: kube-system
+type: kubernetes.io/tls
+data:
+  tls.crt: {{ $caCert }}
+  tls.key: {{ $caKey }}
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/snapshot-validation-webhook/templates/webhook.yaml]: to remove since this secret is not used.

**Review by guangbochen — APPROVED** [review]
LGTM, verified that the webhook pod was running correctly, was able to create and restore the VM or volume snapshots successfully and the invalid snapshot case is rejected by the webhook.

**Comment by guangbochen** [comment]
And please squash the commits before the PR merge, thanks.

**Review by futuretea — APPROVED** [review]
LGTM, thanks. Test Plan Passed

**Review by bk201 — APPROVED** [review]
LGTM!

**Comment by bk201** [comment]
@mergifyio backport v1.2
